sway/lock.c

```diff
@@ -0,0 +1,179 @@
+#define _POSIX_C_SOURCE 200809L
+#include "log.h"
+#include "sway/input/keyboard.h"
+#include "sway/input/seat.h"
+#include "sway/lock.h"
+#include "sway/output.h"
+#include "sway/server.h"
+
+struct sway_session_lock_surface {
+	struct wlr_session_lock_surface_v1 *lock_surface;
+	struct sway_output *output;
+	struct wlr_surface *surface;
+	struct wl_listener map;
+	struct wl_listener destroy;
+	struct wl_listener surface_commit;
+	struct wl_listener output_mode;
+};
+
+static void handle_surface_map(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_surface* surf = wl_container_of(listener, surf, map);
+	sway_force_focus(surf->surface);
+	output_damage_whole(surf->output);
+}
+
+static void handle_surface_commit(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_surface* surf = wl_container_of(listener, surf, surface_commit);
+	output_damage_surface(surf->output, 0, 0, surf->surface, false);
+}
+
+static void handle_output_mode(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_surface* surf = wl_container_of(listener, surf, output_mode);
+	wlr_session_lock_surface_v1_configure(surf->lock_surface, surf->output->width, surf->output->height);
+}
+
+static void handle_surface_destroy(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_surface* surf = wl_container_of(listener, surf, destroy);
+	wl_list_remove(&surf->map.link);
+	wl_list_remove(&surf->destroy.link);
+	wl_list_remove(&surf->surface_commit.link);
+	wl_list_remove(&surf->output_mode.link);
+	output_damage_whole(surf->output);
+	free(surf);
+}
+
+static void handle_new_surface(struct wl_listener *listener, void *data) {
+	struct wlr_session_lock_surface_v1 *lock_surface = data;
+	struct sway_session_lock_surface* surf = calloc(1, sizeof(*surf));
+	if (surf == NULL)
+		return;
+
+	sway_log(SWAY_DEBUG, "new lock layer surface");
+
+	struct sway_output *output = lock_surface->output->data;
+	wlr_session_lock_surface_v1_configure(lock_surface, output->width, output->height);
```

Comment:
This configures the surface with the buffer size. I don't think this will work correctly on transformed or HiDPI outputs.

Reply:
I've tested this with a scale=2 display and it sends the logical size, which matches my read of the protocol ("The width and height are in surface-local coordinates").  Also tested for a transformed surface.

Reply:
Oh, I mixed up `sway_output` and `wlr_output`. Sorry for the noise, your code is correct!